Balance/Login/SignInView.swift

```diff
@@ -232,12 +255,86 @@ struct SignInView: View {
     
     func save(value: ProfileUser) {
         do {
+            if $displayName.wrappedValue.isEmpty ||
+                $email.wrappedValue.isEmpty ||
+                $parentEmail.wrappedValue.isEmpty ||
+                $birthDate.wrappedValue.toString().isEmpty ||
+                $country.name.wrappedValue.isEmpty ||
+                $phone.wrappedValue.isEmpty ||
+                $password.wrappedValue.isEmpty ||
+                $confirmPassword.wrappedValue.isEmpty {
+                self.alertMsg = "All the information is required"
+                showingAlert = true
+                return
+            }
+            
+            if $password.wrappedValue !=
+                $confirmPassword.wrappedValue {
+                self.alertMsg = "The confirmation password mismatch"
+                showingAlert = true
+                return
+            }
+            
+            if !isValidPassword() {
+                self.alertMsg = "Password Error: \n" + getMissingValidation(str: $password.wrappedValue).rawValue
+                showingAlert = true
+                return
+            }
+            
             let data = try JSONEncoder().encode(value)
             UserDefaults.standard.set(data, forKey: "user")
         } catch {
             print("Unable to Encode User (\(error))")
         }
     }
+    
+    func isValidPassword() -> Bool {
+        // least one uppercase,
+        // least one digit
+        // least one lowercase
+        // least one symbol
+        //  min 8 characters total
+        let password = $password.wrappedValue.trimmingCharacters(in: CharacterSet.whitespaces)
+        let passwordRegx = "^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&<>*~:`-]).{8,}$"
+        let passwordCheck = NSPredicate(format: "SELF MATCHES %@", passwordRegx)
+        return passwordCheck.evaluate(with: password)
+    }
+    
+    func getMissingValidation(str: String) -> [String] {
+        var errors: [String] = []
+        if !NSPredicate(format: "SELF MATCHES %@", ".*[A-Z]+.*").evaluate(with: str) {
+            errors.append("At least one uppercase")
+        }
+        
+        if !NSPredicate(format: "SELF MATCHES %@", ".*[0-9]+.*").evaluate(with: str) {
+            errors.append("At least one digit")
+        }
+
+        if !NSPredicate(format: "SELF MATCHES %@", ".*[!&^%$#@()/]+.*").evaluate(with: str) {
+            errors.append("At least one symbol")
+        }
+        
+        if !NSPredicate(format: "SELF MATCHES %@", ".*[a-z]+.*").evaluate(with: str) {
+            errors.append("At least one lowercase")
+        }
+        
+        if str.count < 8 {
+            errors.append("Min 8 characters total")
+        }
+        return errors
+    }
```

Comment:
@gperissetcelteeka just a thought here because I'm not aware of the reach of CardinalKit and/or Firebase but do you know if password policy is something that can get set in the backend side? I'm just concerned about the cases where some policy changes and we need to release a new version of the app. I don't think this is a pressing need but more something to check if we have the time. Thanks!

Reply:
I have to investigate that.